### PR TITLE
Handle Emote Packets

### PR DESCRIPTION
**Pull request by HashimTheArab**
## Introduction
This pull request broadcasts emote packets to players and implements a cooldown with it so someone doesn't dos attack the server with emotes!

### CODE REVIEW

File: src/network/mcpe/handler/InGamePacketHandler.php [hunk 1]
```diff
@@ -888,4 +889,12 @@ public function handleLevelSoundEvent(LevelSoundEventPacket $packet) : bool{
 		 */
 		return true;
 	}
+
+	public function handleEmote(EmotePacket $packet) : bool {
+		if($this->player->getServer()->getTick() - $this->player->lastEmoteTick > 5){
+			$this->player->getWorld()->broadcastPacketToViewers($this->player->getPosition(), $packet);
```
[the remark this answers is not in the document]
Comment (Prim69) [reply src/network/mcpe/handler/InGamePacketHandler.php]: Should be good now

File: src/network/mcpe/handler/InGamePacketHandler.php [hunk 2]
```diff
@@ -888,4 +889,12 @@ public function handleLevelSoundEvent(LevelSoundEventPacket $packet) : bool{
 		 */
 		return true;
 	}
+
+	public function handleEmote(EmotePacket $packet) : bool {
+		if($this->player->getServer()->getTick() - $this->player->lastEmoteTick > 5){
+			$this->player->getWorld()->broadcastPacketToViewers($this->player->getPosition(), EmotePacket::create($packet->getActorRuntimeId(), $packet->getEmoteId(), $packet::FLAG_SERVER));
```
Comment (dktapps) [inline src/network/mcpe/handler/InGamePacketHandler.php:895]:
```suggestion
			$this->player->getWorld()->broadcastPacketToViewers($this->player->getPosition(), EmotePacket::create($packet->getActorRuntimeId(), $packet->getEmoteId(), EmotePacket::FLAG_SERVER));
```
Don't use late static binding. It makes usage search harder and prevents optimisations.

**Comment by dktapps** [comment]
We should also have an event for this.

**Comment by Prim69** [comment]
Anything else?

**Comment by dktapps** [comment]
The build also needs to pass :)

**Comment by Prim69** [comment]
Passed

**Comment by dktapps** [comment]
I had a think about this and I don't like the logic being embedded in the packet handler like this. It should be possible for a plugin to make a player do an emote.

Player::emote(uuid) and NetworkSession::onEmote(player, uuid) should be added.

**Comment by Prim69** [comment]
Would NetworkSession::onEmote send a new emote packet from the given player to the owner of the network session? If that's the case that doesn't sound like the best solution because a new packet would be created for each player that it's broadcasted to

**Comment by dktapps** [comment]
Yes. Stop premature optimising.

**Comment by Prim69** [comment]
What about this?

Player
```php
public function emote(string $emoteId) : void {
	$this->getServer()->broadcastPackets($this->getViewers(), [EmotePacket::create($this->getId(), $emoteId, EmotePacket::FLAG_SERVER)]);
	}
```

because otherwise it would be this

player
```php
public function doEmote(string $emoteId) : void {
	foreach($this->getServer()->getOnlinePlayers() as $player){
		$player->getNetworkSession()->onEmote($this, $emoteId);
	}
}
```
network session
```php
public function onEmote(Player $player, string $emoteId) : void {
	$this->sendDataPacket(EmotePacket::create($this->getId(), $emoteId, EmotePacket::FLAG_SERVER));
}
```

And I don't see NetworkSession::onEmote being used anywhere else so the first one looks like a better solution to me

**Comment by dktapps** [comment]
Stop premature optimising.

**Comment by dktapps** [comment]
Player is not supposed to interact with packets in any way. That's supposed to be handled exclusively by NetworkSession.

Right now the end goal hasn't been reached, but in the future these methods will be on an interface with multiple possible implementations (e.g. for multi version). Broadcasting packets gets in the way of that and it's going to go away in the future.

There are ways that the performance benefits can be retained for such cases, but for something as trivial as this it's premature optimisation - this is not going to be a hot path and nobody will notice either way. Nobody notices the overhead from creating `TextPacket` to broadcast for every player either, and that's a pathway that has an equivalent level of hotness.

**Comment by Prim69** [comment]
Ok that makes sense I didn't see any reason for the second solution before

**Comment by Prim69** [comment]
Everythings set now except my git pull added a bunch of your commits to the pr, should I figure something out for that or will it be fine?

**Comment by dktapps** [comment]
It's a GitHub bug. Those commits aren't actually part of the git diff.

The actual commit diff looks like this: 

image

**Comment by dktapps** [comment]
The event call and rate-limit check should also be done in `Player`. That way the `lastEmoteTick` doesn't need to be public also.

I also prefer `emote()` over `doEmote()` also.

**Comment by Prim69** [comment]
Anything else?

**Pull request merged**